plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
         }
     }
 
+    private static List<Integer> applyProjection(ColumnHandle expectedColumnHandle, ColumnHandle readColumnHandle)
+    {
+        IcebergColumnHandle expectedColumn = (IcebergColumnHandle) expectedColumnHandle;
+        IcebergColumnHandle readColumn = (IcebergColumnHandle) readColumnHandle;
+
+        List<String> expectedDereferences = expectedColumn.getFullPath();
+        List<String> readerDereferences = readColumn.getFullPath();
+        checkState(expectedDereferences.subList(0, readerDereferences.size()).equals(readerDereferences),
+                "Read column path must be a prefix of the Projected column path");
+
+        List<String> path = expectedDereferences.subList(readerDereferences.size(), expectedDereferences.size());
+        ImmutableList.Builder<Integer> dereferenceChain = ImmutableList.builder();
+        ColumnIdentity columnIdentity = readColumn.getColumnIdentity();
+        for (String dereference : path) {
+            List<ColumnIdentity> children = columnIdentity.getChildren();
+            for (int i = 0; i < children.size(); i++) {
+                if (children.get(i).getName().equals(dereference)) {
+                    dereferenceChain.add(i);
+                    columnIdentity = children.get(i);
+                    break;
+                }
+            }
+        }
+
+        return dereferenceChain.build();
+    }
+
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> columns)
```

Comment:
Add note to javadoc in `ReaderColumns` that iceberg uses it now. (or drop that only Hive uses it)